core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java

```diff
@@ -0,0 +1,158 @@
+    @Override
+    public void handleNameResolutionError(Status error) {}
+
+    @Override
+    public void shutdown() {}
+  };
+
+  @VisibleForTesting
+  static final SubchannelPicker BUFFER_PICKER = new SubchannelPicker() {
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return PickResult.withNoResult();
+    }
+
+    @Override
+    public String toString() {
+      return "BUFFER_PICKER";
+    }
+  };
+
+  private final Helper helper;
+  private LoadBalancer currentLb = NOOP_BALANCER;
+  private LoadBalancer pendingLb = NOOP_BALANCER;
+  @Nullable
+  private String currentPolicyName;
+  @Nullable
+  private String pendingPolicyName;
+  private boolean isReady;
+  private ConnectivityState pendingState;
+  private SubchannelPicker pendingPicker;
+
+  public GracefulSwitchLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  /** Gracefully switch to a new load balancing policy. */
+  public void switchTo(LoadBalancerProvider newLbProvider) {
+    checkNotNull(newLbProvider, "newLbProvider");
+
+    String newPolicyName = newLbProvider.getPolicyName();
+    if (newPolicyName.equals(pendingPolicyName)) {
+      return;
+    }
+    pendingLb.shutdown();
+    pendingLb = NOOP_BALANCER;
+    pendingPolicyName = null;
+    pendingState = ConnectivityState.IDLE;
+    pendingPicker = BUFFER_PICKER;
+
+    if (newPolicyName.equals(currentPolicyName)) {
+      return;
+    }
+
+    class PendingHelper extends ForwardingLoadBalancerHelper {
+      LoadBalancer lb;
+
+      @Override
+      protected Helper delegate() {
+        return helper;
+      }
+
+      @Override
+      public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+        if (lb == pendingLb) {
+          checkState(isReady, "there is pending lb while current lb has been out of READY");
+          pendingState = newState;
+          pendingPicker = newPicker;
+          if (newState == ConnectivityState.READY) {
+            swap();
+          }
+        } else if (lb == currentLb) {
+          isReady = newState == ConnectivityState.READY;
+          if (!isReady && pendingLb != NOOP_BALANCER) { // current policy exits READY, swap
+            swap();
+          } else {
+            helper.updateBalancingState(newState, newPicker);
+          }
+        }
+      }
+    }
+
+    PendingHelper pendingHelper = new PendingHelper();
+    pendingHelper.lb = newLbProvider.newLoadBalancer(pendingHelper);
+    pendingLb = pendingHelper.lb;
+    pendingPolicyName = newPolicyName;
+    if (!isReady) { // if the old policy is not READY at the moment, swap to the new one right now
+      swap();
+    }
+  }
+
+  private void swap() {
+    currentLb.shutdown();
+    currentLb = pendingLb;
+    currentPolicyName = pendingPolicyName;
+    pendingLb = NOOP_BALANCER;
+    pendingPolicyName = null;
+    helper.updateBalancingState(pendingState, pendingPicker);
```

Comment:
Let's move the `shutdown()` after this line (alternatively move this before the shutdown).

Reply:
Done.